com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java

```diff
@@ -23,29 +31,166 @@
+			"Cannot initialize VIATRA Query Engine until the project is loaded.";
+	private static final Logger LOGGER = Logger.getLogger(ViatraQueryAdapter.class);
 	
-	private AdvancedViatraQueryEngine engine;
+	private Optional<AdvancedViatraQueryEngine> engine = Optional.empty();
 	private Project project;
 	private boolean engineDisposable = true;
 	private Set<String> identifierMap = new HashSet<>();
+	private boolean isInitialized = false;
+	private Notifier[] notifiers;
+	private List<Consumer<AdvancedViatraQueryEngine>> initializationActions = new ArrayList<>();
 
-	private ViatraQueryAdapter(AdvancedViatraQueryEngine engine, Project project) {
-		this.engine = engine;
+	private ViatraQueryAdapter(Project project, Notifier[] notifiers) {
 		this.project = project;
+		this.notifiers = notifiers;
 	}
-	
+
+	/**
+	 * @deprecated Use {@link #getInitializedEngineChecked()} or {@link #getInitializedEngine()} instead
+	 * @return initialized engine
+	 * @throws ViatraQueryException if no query engine was initialized before and initialization failed
+	 */
+	@Deprecated
 	public AdvancedViatraQueryEngine getEngine() {
+		return getInitializedEngineChecked();
+	}
+
+	/**
+	 * Returns an initialized VIATRA query engine stored in the adapter.
+	 * If a query engine was initialized earlier, that engine will be returned;
+	 * if no query engine was initialized earlier, a new one is created. If
+	 * the initialization fails, e.g. the engine was requested before the
+	 * model was loaded entirely, a {@link ViatraQueryException} is thrown.
+	 * </p>
+	 * This method should only be called if the engine initialization is
+	 * expected to be completed successfully, otherwise it is recommended
+	 * to use {@link #getInitializedEngine()},
+	 * {@link #executeActionOnEngine(Consumer)}
+	 * or {@link #requireQueries(IQuerySpecification)} instead
+	 * as they are aimed to handle the initialization issues better.
+
+	 * @return the initialized engine
+	 * @throws ViatraQueryException if no query engine was initialized before and initialization failed
+	 */
+	public AdvancedViatraQueryEngine getInitializedEngineChecked() {
+		return getInitializedEngine()
+				.orElseThrow(() ->
+					new ViatraQueryException(MESSAGE_ENGINE_NOT_READY, MESSAGE_ENGINE_NOT_READY));
+	}
+
+	/**
+	 * Provide the initialized VIATRA if possible.<br />
+	 * If the engine hasn't been initialized but it can be, then it will be initialize and
+	 * every initialization action registered before will be executed.
+	 * 
+	 * @return the initialized VIATRA engine if possible, otherwise an empty Optional
+	 * @see ViatraQueryAdapter#executeActionOnEngine(Consumer)
+	 * @see ViatraQueryAdapter#requireQueries(IQuerySpecification)
+	 */
+	public Optional<AdvancedViatraQueryEngine> getInitializedEngine() {
+		if(!isInitialized) {
+			initializeEngine();
+		}
+		return engine;
+	}
+	
+	/**
+	 * Checks if an engine is available and execute the given action on it if yes.
```

Comment:
I would reword this as follows:

Executes the give action on the engine associated with the current adapter. If the engine is not yet initialized, the action is saved and will be executed later when the engine gets initialized.

Reply:
I've replaced the previous JavaDoc with your suggestion